File: 670-Maximum-Swap/sam.js
/**
 * @param {number} num
 * @return {number}
 */
var maximumSwap = function (num) {
  const nums = String(num).split("").map(Number);
  const suffix = Array(nums.length).fill(null);
  suffix[suffix.length - 1] = [nums[nums.length - 1], nums.length - 1];

  for (let i = nums.length - 2; i >= 0; i--) {
    if (suffix[i + 1][0] >= nums[i]) {
      suffix[i] = [...suffix[i + 1]];
    } else {
      suffix[i] = [nums[i], i];
    }
  }

  for (let i = 0; i < nums.length; i++) {
    if (suffix[i][0] > nums[i]) {
      const j = suffix[i][1];
      [nums[i], nums[j]] = [nums[j], nums[i]];
      break;
    }
  }

  return Number(nums.join(""));
};

File: 670-Maximum-Swap/sam.py
class Solution:
    def maximumSwap(self, num: int) -> int:
        nums = list(str(num))

        suffix_max = [None] * (len(nums))
        suffix_max[-1] = (nums[-1], len(nums) - 1)

        for i in range(len(nums) - 2, -1, -1):
            if suffix_max[i + 1][0] >= nums[i]:
                suffix_max[i] = (suffix_max[i + 1][0], suffix_max[i + 1][1])
            else:
                suffix_max[i] = (nums[i], i)

        for i, n in enumerate(nums):
            if suffix_max[i][0] > n:
                nums[i], nums[suffix_max[i][1]] = nums[suffix_max[i][1]], nums[i]
                break

        return int("".join(nums))    


